Kernel/SubstHelper.hpp

```diff
@@ -26,6 +26,41 @@ namespace Kernel {
 
 using namespace Lib;
 
+struct SubstApplicator {
+  virtual ~SubstApplicator() = default;
+  virtual TermList operator()(unsigned v) const = 0;
+};
+
+/**
+ * Term with a substitution applied to it lazily.
+ */
+struct AppliedTerm
+{
+  TermList term;
+  bool aboveVar;
+  const SubstApplicator* applicator;
```

Comment:
Could we eliminate `aboveVar` and check whether `applicator == nullptr` instead?

Reply:
Sadly, no as `applicator==nullptr` implies `aboveVar==false` but the converse does not hold.